sources/api/netdog/src/net_config/devices/bonding.rs

```diff
@@ -0,0 +1,184 @@
+use super::{error, Dhcp4ConfigV1, Dhcp6ConfigV1, Result, RouteV1, StaticConfigV1, Validate};
+use super::{validate_addressing, HasDhcp, HasRoutes, HasStatic};
+use crate::interface_name::InterfaceName;
+use serde::Deserialize;
+use snafu::ensure;
+use std::convert::TryFrom;
+use std::net::IpAddr;
+
+#[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct NetBondV1 {
+    pub(crate) primary: Option<bool>,
+    pub(crate) dhcp4: Option<Dhcp4ConfigV1>,
+    pub(crate) dhcp6: Option<Dhcp6ConfigV1>,
+    pub(crate) static4: Option<StaticConfigV1>,
+    pub(crate) static6: Option<StaticConfigV1>,
+    #[serde(rename = "route")]
+    pub(crate) routes: Option<Vec<RouteV1>>,
+    #[allow(dead_code)]
+    kind: BondKind,
+    pub(crate) mode: BondMode,
+    #[serde(rename = "min-links")]
+    pub(crate) min_links: Option<u32>,
+    #[serde(rename = "monitoring")]
+    pub(crate) monitoring_config: BondMonitoringConfig,
+    pub(crate) interfaces: Vec<InterfaceName>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(try_from = "String")]
+pub(crate) struct BondKind(String);
+
+impl TryFrom<String> for BondKind {
+    type Error = error::Error;
+
+    fn try_from(value: String) -> Result<Self> {
+        match value.to_lowercase().as_str() {
+            "bond" => Ok(BondKind(value)),
+            _ => error::InvalidNetConfigSnafu {
+                reason: format!("kind of {} does not match bond", value.as_str()),
+            }
+            .fail(),
+        }
+    }
+}
```

Comment:
Do we really want to treat `Bond` and `boND` as valid? It's not how we do it with other enumerable strings, e.g. in the API, where we require the casing to be correct. Also then we are storing the funkily-cased string for later use.

If `kind` is dead code anyway (only used for validation during deserialization) then presumably there is no reason to store the user's string. Also, if we wanted we could probably do code re-use and have

```suggestion
#[derive(Debug, Deserialize)]
#[serde(try_from = "String")]
pub(crate) struct BondKind;

impl TryFrom<String> for BondKind {
    type Error = error::Error;

    fn try_from(value: String) -> Result<Self> {
        match value.as_str() {
            "bond" => Ok(BondKind),
            _ => error::InvalidNetConfigSnafu {
                reason: format!("kind of '{}' does not match 'bond'", value.as_str()),
            }
            .fail(),
        }
    }
}
```

Also, I'm sad we can't do this... it's probably coming to the language at some point:

```rust
#[derive(Debug, Deserialize)]
#[serde(try_from = "String")]
pub(crate) struct ValidString<const S: &'static str>;

impl<S> TryFrom<String> for ValidString<const S: &'static str>
{
    type Error = error::Error;

    fn try_from(value: String) -> Result<Self> {
        match value.as_str() {
            S => Ok(BondKind),
            _ => error::InvalidNetConfigSnafu {
                reason: format!("kind of '{}' does not match '{}'", value, S),
            }
            .fail(),
        }
    }
}
```

Reply:
For the naming being case sensitive. This is a key we made up (`kind`) only has a helper to both the user and us to make sure we deserialize the intended type of device. Not accepting "Bond" or "BOND" feels slightly more strict than needed. It can be infuriating to lose time to something as silly as a capitalization difference when configuring your system, and since a user has to try a bare metal boot cycle which can take minutes to try things, I have aimed to be as forgiving as possible. I can be talked into keeping the case strict to stay consistent with everything else in the api server, but I really think its a bad customer experience. 

On the storing the string side, I think you are right, I had assumed that `try_from` required you to actually be doing this on a resulting string, but it appears that is not necessary. This is still not ideal, and @zmrow and I have some ideas on cleaning this up further, so I may just take this refactor or do more depending on how much we like this slightly better improvement.

Reply:
I ended up going a completely different way in the logic here. Not sure if its better but would like your take. It feels both better and worse to just implement this logic in the deserialize logic directly rather than use a separate struct and try_from for this.

Reply:
I was able to clean this up even more with @zmrow so I think this turned out nicely now!